### DISCUSSION TITLE
dcp Strange Behavior

### DISCUSSION THREAD

**Discussion by EnsonoStevenPerva** [IBM/zoau-samples#49] [Q&A]
I have run into a bit of odd behavior with `dcp` and bouncing data from a dataset to Unix and back. I know it's not the most practical use-case, but I wanted to raise my experience here. It seems like unless some access has occurred, copying a dataset to USS doesn't seem to update some sort of housekeeping that allows that file to be accessed until it is touched. Here's the problem I'm seeing:

```
decho "Testing" IBMUSER.NEW.DSN
cat "//'IBMUSER.NEW.DSN'"
<observe the data exists>
dcp IBMUSER.NEW.DSN /u/IBMUSER/testfile
dcp /u/IBMUSER/testfile IBMUSER.NEW.COPY
<observe BGYSC1103E error message indicating /u/IBMUSER/testfile doesn't exist>
cat /u/IBMUSER/testfile
dcp /u/IBMUSER/testfile IBMUSER.NEW.COPY
<observe the command complete without error>
```

Is anyone able to reproduce this? `zoaversion` reports `2022/09/07 15:26:50 CUT V1.2.1 d2f6557 1880`

**Comment by EnsonoStevenPerva** [comment]
Also, is this best suited for an "Issue"? Seems like that's more applicable to the repo versus a discussion about ZOAU.

**Reply by EnsonoStevenPerva** [reply]
Additional info: it appears that error message is triggering from something else and the copy is, in fact, being complete. Simply issuing a `cat "//'IBMUSER.NEW.COPY'"` after the error message will show that the copy was complete. Something else is throwing the error message as part of the copy process...

**Comment by AnthonyGiorgio** [comment]
I just tried it on my development system, and I'm not seeing an error message as described.  Could you amend your log to include the exact error message?

**Reply by EnsonoStevenPerva** [reply]
Absolutely! Here's what I get:

```
IBMUSER:/u/IBMUSER: >decho "Testing" IBMUSER.TEST
IBMUSER:/u/IBMUSER: >dcp IBMUSER.TEST /u/IBMUSER/ibmtest
IBMUSER:/u/IBMUSER: >dcp /u/IBMUSER/ibmtest IBMUSER.TEST2
BGYSC1103E No datasets match pattern: /u/IBMUSER/ibmtest.
IBMUSER:/u/IBMUSER: >dls IBMUSER.*
IBMUSER.TEST
IBMUSER.TEST2
```

**Reply by AnthonyGiorgio** [reply]
Message `BGYSC1103E` comes from `dls`, so it looks like that must be called by `dcp` somehow.

**Reply by EnsonoStevenPerva** [reply]
I wonder if there's something treating the path as a dataset name in dls. The results of the error seem to not interfere with the actual copy. The copy works fine despite the error as you can see. Are you able to reproduce this at all?

**Reply by AnthonyGiorgio** [reply]
Why don't you try tweaking `dlshelper` and uncomment the `set -x` line near the top of the file?  Then you can see which of the calls to `mvscmdmsghelper` in the script is printing the message.

I have a feeling that we might have a race condition here.

**Reply by EnsonoStevenPerva** [reply]
Here's the result of that test:
```
IBMUSER:/u/IBMUSER: >decho "Testing" IBMUSER.REPRO.D110222
IBMUSER:/u/IBMUSER: >dcp IBMUSER.REPRO.D110222 /u/IBMUSER/repro
IBMUSER:/u/IBMUSER: >dcp /u/IBMUSER/repro IBMUSER.REPRO.D110222.TWO
+ set -o noglob
+ InvalidOption=1
+ TooFewParms=2
+ NoDatasetsMatch=3
+ NoDetailsForMigratedDatasets=4
+ details=0
+ size=0
+ blksize=0
+ lastreftime=0
+ migrated=0
+ debug=0
+ verbose=0
+ quiet=0
+ opts=
+ getopts :bsluvdmhq opt
+ quiet=1
+ getopts :bsluvdmhq opt
+ [ 1 -eq 1 ]
+ details=0
+ [ 0 -eq 1 ]
+ expr 2 - 1
+ 1<TMP> /tmp/shfDbfhHhbA
+ shift 1
+ [ 1 -lt 1 ]
+ + ListSummary IBMUSER.REPRO.D110222.TWO
+ 1<TMP> /tmp/shdgehGEccA
+ MAXRC=0
+ + dlsrawhelper IBMUSER.REPRO.D110222.TWO
+ 1<TMP> /tmp/shbfcgABFGA
datasets=
+ RC=1
+ [ 1 -gt 0 ]
+ MAXRC=1
+ echo
+ exit 1
datasets=
+ RC=1
+ [ 1 -gt 4 ]
+ [  =  ]
+ [ 1 -eq 0 ]
+ exit 1
+ set -o noglob
+ InvalidOption=1
+ TooFewParms=2
+ NoDatasetsMatch=3
+ NoDetailsForMigratedDatasets=4
+ details=0
+ size=0
+ blksize=0
+ lastreftime=0
+ migrated=0
+ debug=0
+ verbose=0
+ quiet=0
+ opts=
+ getopts :bsluvdmhq opt
+ details=1
+ getopts :bsluvdmhq opt
+ blksize=1
+ details=1
+ getopts :bsluvdmhq opt
+ size=1
+ details=1
+ getopts :bsluvdmhq opt
+ [ 0 -eq 1 ]
+ [ 0 -eq 1 ]
+ expr 2 - 1
+ 1<TMP> /tmp/shgCCeAHBgA
+ shift 1
+ [ 1 -lt 1 ]
+ + ListSummary /u/IBMUSER/repro
+ 1<TMP> /tmp/shdbDEdabaA
+ MAXRC=0
+ + dlsrawhelper /u/IBMUSER/repro
+ 1<TMP> /tmp/shHeaeDCAcA
datasets=
+ RC=1
+ [ 1 -gt 0 ]
+ MAXRC=1
+ echo
+ exit 1
datasets=
+ RC=1
+ [ 1 -gt 4 ]
+ [  =  ]
+ [ 0 -eq 0 ]
+ mvscmdmsghelper dls 3 E /u/IBMUSER/repro
BGYSC1103E No datasets match pattern: /u/IBMUSER/repro.
+ exit 1
IBMUSER:/u/IBMUSER: >
```
